### ISSUE TITLE
lane guidance missing when navigating an imported GPX track

### ISSUE THREAD

**Issue by TheSnoozer** [osmandapp/OsmAnd#14832]
Hello thanks for the great app!
One small thing (I haven't found an existing bug that describes the problem) I noticed is that 
the lane guidance is missing when navigating an imported GPX track.

As a simple expample take any route in google (https://goo.gl/maps/QfrJDjRdP1Ugc3MP9) and convert it to a GPX track (https://mapstogpx.com).
Or download the track here:
[mapstogpx20220712_134524.zip](https://github.com/osmandapp/OsmAnd/files/9093542/mapstogpx20220712_134524.zip)

I have tried other GPX tracks and none of them indicate any lane guidance.

# 🐞 routing report

### Routing engine

- [X] OsmAnd's in-app offline routing with a GPX track
- [ ] Any online routing provider (YOURS, openrouteservice, OSRM, etc.)

### Routing Profile

The car routing profile was used (default settings)

### Actual and expected routes

Lane guidance is displayed when following a GPX track

### Is this a regression?

I do not think this is a regression.

## 🌍 Your Environment

**OsmAnd Version:**

OsmAnd+ 4.2.5

**Comment by scaidermern** [comment]
You have to attach the track to the road first.

Try the following: Go to My Places / Tracks / <your track> / Options / Edit track / Options / Attach to the roads.

Here, select a navigation profile. Then wait for the tiny blue progress bar (between the map and the bottom widget) to complete. Now press apply and save the track (overwrite the current track or save it as a new track).

**Comment by TheSnoozer** [comment]
Thank you for your fast response. This option is certainly hidden a bit too much. Given that it sounds like I want I'm going ahead and close the issue.

PS: I don't know if it actually works my potato phone crashes with a `OutOfMemoryError` when trying to calculate a relative short GPX track (40km).

**Comment by scaidermern** [comment]
> Thank you for your fast response. This option is certainly hidden a bit too much.

Indeed, I fully agree here. Unfortunately I have no good idea how to make this feature more obvious. A toast or a dialog when trying to use a regular GPX track for navigation?

**Comment by TheSnoozer** [comment]
> Indeed, I fully agree here. Unfortunately I have no good idea how to make this feature more obvious. A toast or a dialog when trying to use a regular GPX track for navigation?

Thanks for your input and I'm glad you also share the impression that this setting is buried too deep.
How to make it more visible is certainly a really good question. My personal main reasons to use OsmAnd is the offline navigation, the lane guidance and the fact that it's open source. So right from the beginning my expectation in "following" or "navigating" a track is that I have the lane guidance avilable ;-) 

I don't know what your option enables or what data is additionally saved with the track, but for me a GPX track is effectively just a set of waypoints. 
Based on your description I'm assuming that the option above somehow adds more data to the track. If so I would claim that Osm should be able to detect if the data is there or not. If no such data is available simply prompt the user in the navigation dialog when "navigating" the route if Osm should try to calculate a route that attaches to the selected GPX track. Clearly the user should be made more aware that a "raw" GPX track might lack some navigational information (speed limits, lane guidance, etc...).

TLDR: To make this more visible I think Osm simply would need to add an additional option / prompt when "navigating" a GPX track.

**Comment by mikehgentry** [comment]
It also wouldn't do any harm in my opinion having a link to the 'Track options' screen in

Navigation -> Some track -> Options

(especially since you already have a link to 'Navigation settings' in there). Presumably you could just grey it out with no track selected?

**Comment by scaidermern** [comment]
@victorosm Any opinion from you on making this feature more obvious to users? This question/problem comes up rather often.

**Comment by vshcherb** [comment]
![Uploading telegram-cloud-photo-size-2-5366173625167757262-x.jpg…]()

**Comment by vshcherb** [comment]
https://github.com/osmandapp/OsmAnd/issues/14855

**Comment by Flightkick** [comment]
Found this report because I experienced the same issue, frustrating to discover that lane indication isn't working when you're already on the road. Don't get me wrong, I'm not blaming anyone and I'm very grateful for this software being available as FLOSS. Just pointing out that it might be more pleasant for users to be aware of this beforehand.

That's a nice addition @vshcherb! Unfortunately, it was not shown to me during the flow of importing the GPX and starting navigation. There is a button for 'Attach to the roads' if you bother going to the route options, but that description is absent there. I figured I probably shouldn't mess around with that so that I wouldn't accidentally mess up the navigation.

<img src="https://github.com/osmandapp/OsmAnd/assets/8693699/98deeb9f-a91f-403a-8fc7-0cbfa48051f8" width="400">

Under 'Configure Screeen' > 'Top panel', I got 'Lanes' active. Which makes the lanes widget show up, maybe that widget could also somehow make the user aware that lane indication is not available and direct the user on how to resolve that.

In any case, thanks for reporting and adding the feature. At least now I know that tracks should be attached to roads the next time.

**Comment by vshcherb** [comment]
There is attach to the roads message box inside Navigation. If you start Navigate by GPX it will be suggested to use Attach to the roads as well

**Comment by nikos2wheels** [comment]
I can't get the lane guidance to display on a GPX track. When I plan a route under Options I see Attach to Roads and I click on it but nothing happens. When I try to navigate with GPX track the lane guidance is missing. Was this issue resolved?

Also when I load the track, under Navigation Settings there is Detailed track guidance.  (Osmand will match your track to nearby roads for enhanced turn-by-turn guidance) Is this the same as Attach to Roads? I tried both Ask Every time and Always but that didn't show Lane Guidance. When I chose Ask Every time I was never asked, how is it meant to ask me?

**Comment by scaidermern** [comment]
> When I plan a route under Options I see Attach to Roads and I click on it but nothing happens

How long did you wait? Depending on the length of the track OsmAnd takes a while to attach it to the map data.

**Comment by nikos2wheels** [comment]
So I tried to edit the track and choose attach to the roads and it does show me the lane guidance.

However, in the profile under Navigation Settings there is Detailed track guidance. (Osmand will match your track to nearby roads for enhanced turn-by-turn guidance). When I choose Always it fails to show me the Lane Guidance. Isn't this option meant to always attach track to roads? If yes, then it fails. I am using OsmAND 4.9 on Android.
